src/hotspot/share/gc/shenandoah/shenandoahMarkClosures.cpp

```diff
@@ -60,21 +60,6 @@ void ShenandoahFinalMarkUpdateRegionStateClosure::heap_region_do(ShenandoahHeapR
       }
     }
 
-    if (ShenandoahHeap::heap()->mode()->is_generational()) {
-      // Allocations move the watermark when top moves, however compacting
-      // objects will sometimes lower top beneath the watermark, after which,
-      // attempts to read the watermark will assert out (watermark should not be
-      // higher than top). The right way™ to check for new allocations is to compare
-      // top with the TAMS as is done earlier in this function.
-      if (top > tams) {
-        // There have been allocations in this region since the start of the cycle.
-        // Any objects new to this region must not assimilate elevated age.
-        r->reset_age();
-      } else if (ShenandoahHeap::heap()->is_aging_cycle()) {
-        r->increment_age();
```

Comment:
Hmm, we used to increment age for every region during final mark, but now we are doing it during final update refs. This means no regions will be aged for cycles that skip evacuation. Is that what we want? this will have the effect of lowering the promotion rate.

Reply:
This is a good point.  I guess we can increment age in final_mark for any region that is not placed into the collection set.  In the case that there is sufficient immediate trash that we do not have to evacuate any regions, then all region ages can be incremented.  That would be better.  I'll work on this change.

Reply:
It's a bit more complicated than at first glance.  We don't want to increment age before we choose the collection set because that would influence what goes into the collection set.  Also, we don't want the age of the region to be added to the object age when deciding whether to promote until after all the objects in this region have not been evacuaetd during an aging cycle.  I've made a change that implements the desired behavior.  It is basically some special fixup in the case that we short-circuit evacuation.  What do you think?